apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java

```diff
@@ -234,14 +233,31 @@ void onConnectionError() {
             } catch (Exception e) {
                 logger.debug("Exception while interacting with APM Server", e);
             } finally {
-                if (connection != null) {
-                    connection.disconnect();
-                }
+                closeInputStream(connection);
             }
         }
         return results;
     }
 
+    private void closeInputStream(@Nullable HttpURLConnection connection) {
+        if (connection != null) {
+            try {
+                int responseCode = connection.getResponseCode();
+                InputStream responseInputStream = null;
+                if (responseCode < 400) {
```

Comment:
rather than checking for the response code, check if `getErrorStream` returns a non-null result.

Reply:
changed